src/accelerate/test_utils/scripts/external_deps/test_ds_multiple_model.py

```diff
@@ -0,0 +1,328 @@
+
+    def collate_fn(examples):
+        # On TPU it's best to pad everything to the same length or training will be very slow.
+        if accelerator.distributed_type == DistributedType.XLA:
+            return tokenizer.pad(examples, padding="max_length", max_length=128, return_tensors="pt")
+        return tokenizer.pad(examples, padding="longest", return_tensors="pt")
+
+    # Instantiate dataloaders.
+    train_dataloader = DataLoader(
+        tokenized_datasets["train"], shuffle=True, collate_fn=collate_fn, batch_size=batch_size
+    )
+    eval_dataloader = DataLoader(
+        tokenized_datasets["validation"], shuffle=False, collate_fn=collate_fn, batch_size=EVAL_BATCH_SIZE
+    )
+
+    return train_dataloader, eval_dataloader
+
+
+test_file_path = __file__
+path = Path(test_file_path).resolve()
+test_file_dir_str = str(path.parent.parent.parent.parent.parent.parent)
+
+# Create our DS plugins
+# We use custom schedulers and optimizers, hence `model_only`
+ds_config_file = dict(
+    zero2=f"{test_file_dir_str}/tests/deepspeed/ds_config_zero2_model_only.json",
+    zero3=f"{test_file_dir_str}/tests/deepspeed/ds_config_zero3_model_only.json",
+)
+
+
+def single_model_training(config, args):
+    # Training a single model, we have a `noise` model that is untrainable used to inject some noise into the training process
+    num_epochs = config["num_epochs"]
+    zero2_plugin = DeepSpeedPlugin(hf_ds_config=ds_config_file["zero2"])
+    zero3_plugin = DeepSpeedPlugin(hf_ds_config=ds_config_file["zero3"])
+
+    deepspeed_plugins = {"training": zero2_plugin, "inference": zero3_plugin}
+
+    # Initialize accelerator
+    accelerator = Accelerator(
+        deepspeed_plugin=deepspeed_plugins,
+        mixed_precision="bf16",
+    )
+
+    # Initialize model under zero2 plugin
+    assert get_active_deepspeed_plugin(accelerator.state) is zero2_plugin
+    trained_model = AutoModelForSequenceClassification.from_pretrained(args.model_name_or_path)
+    train_dataloader, eval_dataloader = get_dataloaders(
+        accelerator, batch_size=config["batch_size"], model_name=args.model_name_or_path
+    )
+    max_training_steps = len(train_dataloader) * config["num_epochs"]
+    optimizer = AdamW(trained_model.parameters(), lr=config["lr"])
+    lr_scheduler = get_linear_schedule_with_warmup(
+        optimizer, num_warmup_steps=0, num_training_steps=max_training_steps
+    )
+
+    train_dataloader, eval_dataloader, model, optimizer, lr_scheduler = accelerator.prepare(
+        train_dataloader, eval_dataloader, trained_model, optimizer, lr_scheduler
+    )
+
+    # Now prepare the model under zero3 plugin
+    accelerator.state.enable_deepspeed_plugin("inference")
+    assert get_active_deepspeed_plugin(accelerator.state) is zero3_plugin
+    inference_model = NoiseModel()
+    inference_model = accelerator.prepare(inference_model)
+    inference_model.eval()
+
+    # Run training loop
+    accelerator.state.enable_deepspeed_plugin("training")
```

Comment:
On the accelerate side, not particularly much/anything at all, since it's mostly for `deepspeed.init()`. But for the transformers side eventually I believe this should run into some errors between zero2 and zero3, since with zero3 we need to gather params, and this sets a flag that tells transformers we're using zero3 right now.